client/src/connection/rest/index.ts

```diff
@@ -13,263 +13,271 @@ import { ComputeState, getApiConfig } from "./common";
+    //Grab the sessionId
+    const sessionId: string = await getContextValue("SAS.sessionId");
 
-  //Clear out the logs
-  await retLog;
+    if (sessionId === undefined) {
+      //No sessionId in the cache means nothing to reconnect to
+      return undefined;
+    }
 
-  //Now get the results
-  const results = state === ComputeState.Error ? [] : await job.results();
+    //At this point a sessionId was retrieved, so try and re-connect
 
-  const res: RunResult = {
-    html5: "",
-    title: "",
-  };
+    if (this._config.serverId) {
+      const computeServer = new ComputeServer(this._config.serverId);
 
-  /*
-    We can return more than just html, but for right now we are only returning
-    the last HTML file that we get.
-    The last one is returned so that the one created from the vscode injected ods statement is
-    always returned.
-  */
-  for (const result of results.reverse()) {
-    const link = result.links[0];
-    if (link?.type === "text/html") {
-      const html5 = (await job.requestLink<string>(link)).data;
-
-      //Make sure that the html has a valid body
-      if (html5.search('<*id="IDX*.+">') !== -1) {
-        res.html5 = html5;
-        res.title = result.name;
+      try {
+        session = await computeServer.getSession(sessionId);
+      } catch (error) {
+        console.log(
+          `Attempt to reconnect to session ${sessionId} failed. A new session will be started`
+        );
+      }
+    } else {
+      const apiConfig = getApiConfig();
+      const sessions = SessionsApi(apiConfig);
+
+      try {
+        const mySession = (await sessions.getSession({ sessionId: sessionId }))
+          .data;
+        session = ComputeSession.fromInterface(mySession);
+      } catch (error) {
+        console.log(
+          `Attempt to reconnect to session ${sessionId} failed. A new session will be started`
+        );
       }
-
-      break;
     }
-  }
 
-  return res;
-}
+    if (session === undefined) {
+      //If we tried to reconnect and failed, set the cached sessionId to undefined
+      setContextValue("SAS.sessionId", undefined);
+    }
 
-function sessionId() {
-  return computeSession && computeSession.sessionId;
+    return session;
+  };
 }
 
-async function close() {
-  if (sessionId()) {
-    computeSession.delete();
-    computeSession = undefined;
-
-    //Since the session is being closed, remove the cached session id
-    setContextValue("SAS.sessionId", undefined);
+/*
+Prints the job log in an async manner.
+*/
+async function printLog(job: ComputeJob, onLog?: (logs: LogLine[]) => void) {
+  const logs = await job.getLogStream();
+  for await (const log of logs) {
+    onLog(log);
   }
 }
 
 export function getSession(c: Config): Session {
-  config = c;
-  getApiConfig().basePath = config.endpoint + "/compute";
-
-  return {
-    setup,
-    run,
-    cancel,
-    close,
-    sessionId,
-  };
+  getApiConfig().basePath = c.endpoint + "/compute";
+
+  return new RestSession(c);
```

Comment:
`getSession` can be called many times from anywhere. It would be better to make it singleton. Otherwise different components may operate on different instances and may increase complexity.

Reply:
Good catch, I'd done this on the ssh support but forgot to implement it here. See updated.